Makes labor camp sustenance vendor cost labor points

Sealed101 opened the pull request:
## About The Pull Request

On the tin.
Labor camp vendor is a subtype of the sustenance vendor, same stock, different currency.
If a user doesn't have a prisoner ID, they won't be able to buy anything from that vendor.
![изображение](https://github.com/tgstation/tgstation/assets/75863639/e266e168-0c3c-4caf-ab05-879c4a18323c)

## Why It's Good For The Game

Brings to a close the inconsistency of permabrig vendor costing money but the labor camp vendor being free...
...with the caveat that now it actually costs your freedom points. Which has been agreed upon to be equal parts fun and evil.
Closes #40889 
![изображение](https://github.com/tgstation/tgstation/assets/75863639/a10241ca-f7e4-4a62-a945-569c2f8be534)
Keeping it as a fix but lmk

## Changelog

:cl:
fix: Labor Camp Sustenance vendor is no longer free; instead, it takes your labor points that you'd otherwise use for paying off your point goal.
/:cl:

Sealed101 committed:
labor camp vendor subtype
replaces the one in lavaland labor camp with this one
both permabrig and labor camp sustenance vendors cost money now

BlueMemesauce commented:
Don't you get your id taken when sent to labor camp?

Sealed101 commented:
the prisoner ID starts unassigned but the prisoner can assign their bank account to it immediately, which will receive the paycheck still

Sealed101 commented:
waiting on arcane's certified capitalist opinion


https://github.com/tgstation/tgstation/assets/75863639/65ad387d-0053-4f5c-b4fe-75a58bbc0415

TheVekter commented:
Would it be possible to make the vendor use mining points instead of credits? That fixes the issue of players not knowing how to link their bank account to a new ID and makes it possible to earn enough to get a meal if they need one while in gulag.

ArcaneMusic commented:
I talked for awhile about this on the discord but in short: I don't mind if people want to change the vending machines to act like the mining vendor where you get paid in gulag points. I'd go as far as to say that could be kinda fun. I certainly wouldn't stop anyone if they want to go through with making that.

I will say, if it's otherwise going to remain a vending machine, and that vending machine is on station, following the design of other economy related content, I would need the vending machine to cost credits.

Sealed101 committed:
i got it to work but not really
tgui still thinks it's in credits
normal ids break it since they don't have labor points
the way i do it is hacky like all things i do on this repo
send help

Sealed101 commented:
i got the vendor to accept GPP (good prisoner points) but i still need to cleanly resolve interaction of non-prisoners with the vendor, as well as probably prohibit account linking to the prisoner ID or something. also would need to get the vendor to actually display a different name for the currency change

Sealed101 committed:
ok how do i make that string show up
displayed_currency_name doesn't show up for the purchase button, i got it to work with the balance display but here it doesn't work. send help

Sealed101 committed:
ah right it's passed in static data

Sealed101 committed:
vendor screams at you if id isn't a prisoner id
also hopefully properly runs prettier on vending.tsx

Sealed101 committed:
icebox labor camp vendor

Sealed101 committed:
labor camp vendor icon

Sealed101 committed:
Merge branch 'master' into labor-camp-sustenance-betwixt-thee

Sealed101 committed:
hello merge conflict

Sealed101 commented:
it works ™️ 
code's definitely shit and needs further optimizations, and there isn't any logging of the proceedings, but otherwise yeah it works ™️

ArcaneMusic commented:
Ready to undraft for a review orrr

Sealed101 commented:
i mean sure

ArcaneMusic commented on code/modules/vending/sustenance.dm:
```diff
@@ -29,3 +29,67 @@
+	if(LAZYLEN(product_to_vend.returned_products))
+		price_to_use = 0 //returned items are free
+	if(price_to_use && !(paying_scum_id.points >= price_to_use)) //not enough good prisoner points
+		speak("You do not possess the funds to purchase [product_to_vend.name].")
```
Since we're overriding the base context from vending machines to begin with, reword "funds" to something referencing mining points to avoid confusion like you've done when using standard IDs

ArcaneMusic commented on code/modules/vending/sustenance.dm:
```diff
@@ -29,3 +29,67 @@
+	.["extended_inventory"] = extended_inventory
+
+/obj/machinery/vending/sustenance/proceed_payment(obj/item/card/id/paying_id_card, datum/data/vending_product/product_to_vend, price_to_use)
+	if(!istype(paying_id_card, /obj/item/card/id/advanced/prisoner))
```
Any specific reason we want to make it so prisoners are the only ones who can purchase from these vendors? I imagine needing mining points and being in a high security part of prison is exclusive enough.

Sealed101 replied:
this is using labor points that are exclusive to prisoner ID cards. if you want to use the actual mining points that the miners use, I could look into that

Sealed101 replied:
though that would kill the funny aspect of paying your bailout cash for food in gulag

ArcaneMusic replied:
Alright whatever then

ArcaneMusic commented on code/modules/vending/sustenance.dm:
```diff
@@ -29,3 +29,67 @@
+/obj/machinery/vending/sustenance/interact(mob/user)
+	if(isliving(user))
+		var/mob/living/living_user = user
+		if(!(machine_stat & NOPOWER) && !istype(living_user.get_idcard(TRUE), /obj/item/card/id/advanced/prisoner))
```
As above, any reason these HAVE to be prisoner IDs?

ArcaneMusic reviewed (changes requested):
Minor tweaks

Sealed101 committed:
less copypastey ui_data build for funds
fetch_balance_to_use proc handles that, defaulting to returning the card's account balance
on the labor vendor it returns points, and a balance of null if you swapped IDs with the vendor's UI open
also fixes proceed_payment implementation; it woulda just returned and continued to the vending process despite any potential returns that shouldn't let the machine vend

Sealed101 committed:
dmdoc and varnames for _vending.dm
some minor optimizations may be present, but don't count on it

Sealed101 committed:
Merge branch 'master' into labor-camp-sustenance-betwixt-thee

ArcaneMusic reviewed (approved):
Seems fine then
